### DISCUSSION TITLE
Initial State

### DISCUSSION THREAD

**Discussion by Yangsjunqi** [python-control/python-control#753] [Q&A]
Hi,
I was trying to test a simple feedback system with PID controller . Simulation can be easily done with the code below. However, I cant figure out how to specify the initial value (e.g. 10...) of response instead of 0 by default. Would you please tell the solution?
Thanks,
Junqi

```
def sys1(Kp,Ki,Kd):
    G = control.tf([1],[1,1])
    K = control.tf([Kd, Kp,Ki], [1,0])
    S = control.feedback(G*K, 1)
    d = np.linspace(0, 10, 100)
    t, rec = control.initial_response(S, d)
    return rec
```

**Comment by murrayrm** [comment]
If you would like the initial condition response, you need to provide the initial state from which you would like to simulate the system as an argument to `control.initial_response`.  You can do this by providing a third argument to `initial response`:
```
t, y = control.initial_response(sys, T, X0)
```
Note, however, that since you are dealing with transfer functions, the "states" of your system are not well-defined.  The state space representation of your system has three states (1 from the PID controller `K` and two from the system `G`), but the state space realization of `S` is not well-defined.

A more common response for a transfer function is the step response, which you can obtain using `control.step_response`.

If you really want the initial condition response, change the next to last line of your function to be something like
```
t, rec = control.initial_response(S, d, [10, 10, 10])
```

**Reply by Yangsjunqi** [reply]
Thanks for the quick response!
Actually I tried control.step_response and initial_response before as you advised, but didnt get the results I want. The observation is, when it is a open-loop system, the X0 argument works, but when it involves controller as a close-loop, things changed.

**Comment by Dimethylebutane** [comment]
Hi,
I have the exact same issue, simulating closed loop system. The X0 parameters changes nothing if not in the order of 10**8.
I found that the initial value of my system is X0 * 1.4/(2*10^8) (and it is not exactly that ?).
Also my system is non-linear (I reused the example _"Simulating interconnections of systems"_ to connect a discrete controller to a continuous system), and I need to set `X0 = [initVal * 2*10**8 / 1.14, initVal * 2*10**8 / 1.14]`, both states to the same value.

I have very little knowledge about simulation, and I really don't get why this behavior. If someone with more experience than me could explain how to set this X0 value.

Thanks !

**Comment by sawyerbfuller** [comment]
@Dimethylebutane glad to hear that the simulation example for discrete-continuous systems was useful! 

you usually don't want to use nonzero initial conditions. usually you are interested in driving your system, starting from zero, to a new desired position. or simulating what happens when a disturbance enters right before the plant. both of those involve zero initial conditions. 

that said, if all of your systems are state space, then their states are ordered in the order you passed them into`interconnect`. I'm not sure why your states need to be so big - they should match the states of the original system, if it was originally specified in state-space format. if any of your systems were transfer functions, then this will not work (see richard's comment above). maybe try a simpler system to get a feel for how your states work when specifying initial conditions.
